### PR TITLE
DiscreteUniform distribution and example of discrete step method usage.

### PR DESCRIPTION
**Pull request by fonnesbeck**
Also improved metropolis sampler to accept list of mixed-type variables. Miscellaneous bug fixes too.

### CODE REVIEW

File: pymc/distributions/discrete.py [hunk 1]
```diff
@@ -184,6 +182,41 @@ def logp(value):
 
 
 @tensordist(discrete)
+def DiscreteUniform(lower, upper):
```
Comment (jsalvatier) [inline pymc/distributions/discrete.py:185]: Discrete uniform doesn't have mean, median or mode that makes sense as a starting point. There are probably some other distributions that have this problem too. Thoughts on what we should do about this?

Options:
- Add a 'reference_value' on the test value search path
- Round the values on the search path for discrete variables
- ...

File: pymc/distributions/discrete.py [hunk 2]
```diff
@@ -184,6 +182,41 @@ def logp(value):
...
+    logp.__doc__ = """
+        Discrete uniform log-likelihood with bounds ({0},{1}).
+
+        Parameters
+        ----------
+        x : int
+            :math:`lower \leq x \leq upper`
+        """.format(upper, lower)
+
+    mode = floor((upper - lower) / 2.).astype('int32')
```
Comment (jsalvatier) [inline pymc/distributions/discrete.py:216]: Oh, haha, I didn't see you had added this.

File: pymc/distributions/discrete.py [hunk 3]
```diff
@@ -184,6 +182,41 @@ def logp(value):
...
+    :Parameters:
+      - `lower` : Lower limit.
+      - `upper` : Upper limit (upper > lower).
+
+    """
+
+    def logp(value):
+
+        return bound(
+            -log(upper-lower+1),
```
Comment (jsalvatier) [inline pymc/distributions/discrete.py:202]: Should we cast `upper` and `lower` to ints or throw an error if they're floats? If they're floats I think you can run into situations where there's less than 1 total probability.

File: pymc/step_methods/metropolis.py [hunk 4]
```diff
@@ -93,8 +88,8 @@ def astep(self, q0, logp):
         delta = self.proposal_dist() * self.scaling
 
         q = q0 + delta
-        if self.discrete:
-            q = round(q, 0).astype(int)
```
Comment (jsalvatier) [inline pymc/step_methods/metropolis.py:97]: Clever! I like this.

**Comment by jsalvatier** [comment]
Looks good! 

All that's left is:
- Rebase
- Add DiscreteUniform to the distribution tests
- Re-add step method tests for Slice. I removed it since it was causing the tests to fail. Is there a way to mark generated tests as knownfailures? Maybe that's a good reason to not use generated tests too much.

File: pymc/distributions/discrete.py (hunk 1 above)
Comment (fonnesbeck) [inline pymc/distributions/discrete.py:185]: `floor((upper - lower)/2.)`

**Comment by fonnesbeck** [comment]
Taken care of the distribution test in the following commit, and have rebased. The disaster_model example uses a Slice sampler, though I'd like to incorporate more granular tests at some point.

**Pull request merged**